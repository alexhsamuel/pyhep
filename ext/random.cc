//----------------------------------------------------------------------
//
// random.cc
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* Random number generation functions.  */

//----------------------------------------------------------------------
// includes
//----------------------------------------------------------------------

#include <cassert>
#include <cstddef>
#include <sys/time.h>

#include "random.hh"

//----------------------------------------------------------------------
// constants
//----------------------------------------------------------------------

namespace {

// Constants for the first congruential generator.
const static long m1 = 2147483563;
const static long a1 = 40014;
const static long q1 = 53668;
const static long r1 = 12211;

// Constants for the second congruential generator.
const static long m2 = 2147483399;
const static long a2 = 40692;
const static long q2 = 52774;
const static long r2 = 3791;

//----------------------------------------------------------------------
// helper functions
//----------------------------------------------------------------------

inline float
min(float v1,
    float v2)
{
  return (v1 < v2) ? v1 : v2;
}


}  // anonymous namespace

//----------------------------------------------------------------------
// methods
//----------------------------------------------------------------------

ShuffledLEcuyerRandom::ShuffledLEcuyerRandom(long seed)
{
  initialize(seed);
}


void
ShuffledLEcuyerRandom::initialize(long seed)
{
  // If no seed was specified, construct one from the system time.
  if (seed == 0) {
    struct timeval now;
    int result = gettimeofday(&now, NULL);
    assert(result == 0);
    seed = now.tv_sec ^ now.tv_usec;
  }

  // Initialize.
  assert(seed > 0);
  state1_ = seed;
  state2_ = seed;

  // Load the shuffle table (after eight warm-ups).
  for (long j = table_size_ + 7; j >= 0; j--) {
    long k = state1_ / q1;
    state1_ = a1 * (state1_ - k * q1) - k * r1;
    if (state1_ < 0)
      state1_ += m1;
    if (j < table_size_)
      shuffle_table_[j] = state1_;
  }
  last_ = shuffle_table_[0];
}


float
ShuffledLEcuyerRandom::random()
{
  long k;
  static float inv_m1 = 1.0 / m1;

  // Next value from the first congruential generator.
  k = state1_ / q1;
  state1_ = a1 * (state1_ - k * q1) - k * r1;
  if (state1_ < 0)
    state1_ += m1;

  // Next value from the second congruential generator.
  k = state2_ / q2;
  state2_ = a2 * (state2_ - k * q2) - k * r2;
  if (state2_ < 0)
    state2_ += m2;

  // Combine and shuffle.
  int j = last_ / (1 + (m1 - 1) / table_size_);
  last_ = shuffle_table_[j] - state2_;
  shuffle_table_[j] = state1_;
  if (last_ < 1)
    last_ += m1 - 1;

  // Make sure we never return 1.0.
  return min(last_ * inv_m1, 1.0 - 1.2e-7);
}


