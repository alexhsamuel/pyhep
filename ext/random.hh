//----------------------------------------------------------------------
//
// random.hh
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* Random number generation functions.  */

#ifndef __RANDOM_HH__
#define __RANDOM_HH__

//----------------------------------------------------------------------
// classes
//------------------------------------------------------------------------

/* Shuffled L'Ecuyer random number generator.

   Generator for uniform deviates based on 

     L'Ecuyer, P. 1988, Communications of the ACM, vol. 31, pp. 742-774

   as described and implemented as the 'ran2' function in

     Press, William H. et al., Numerical Recipes in C, Cambridge
     University Press, 1988, p. 282.

   This generator is based on two multiplicative congruential generators
   with factors 'a1' and 'a2' and moduli 'm1' and 'm2, with
   multiplication performed with Schrage's algorith.  The values from
   the two sequences are combined, and the sequence is shuffled using a
   table of 'tab_size'.  

   The period of the generator is about 2.3 x 10^18.

*/
class ShuffledLEcuyerRandom
{
public:

  /* Create a new generator.

     'seed' -- A positive integer seed value.  If zero, a seed is chosen
     using the system time. 
  */
  ShuffledLEcuyerRandom(long seed=0);
  
  /* Initialize the generator. 

     'seed' -- A positive integer seed value.  If zero, a seed is chosen
     using the system time. 
  */
  void initialize(long seed=0);

  /* Return a uniform deviate between 0 inclusive and 1 exclusive.  */
  float random();

private:

  // Forbid copies.
  ShuffledLEcuyerRandom(const ShuffledLEcuyerRandom&);

  // Shuffle table size.
  const static int table_size_ = 32;

  // State variables for the congruential generators.
  long state1_;
  long state2_;

  // State of the combined generator; used for shuffling.
  long last_;

  // The shuffle table.
  long shuffle_table_[table_size_];

};

//----------------------------------------------------------------------

#endif  // #ifndef __RANDOM_HH__
