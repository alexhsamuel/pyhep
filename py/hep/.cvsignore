*.pyc
*.pyo
config.py
*-install
pdt.pickle
