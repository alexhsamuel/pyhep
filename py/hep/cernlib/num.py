#-----------------------------------------------------------------------
#
# num.py
#
# Copyright (C) 2004 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Numerical functions based on CERNLIB."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

import ext

#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def chiSquareCDF(x, n):
    """CDF for chi-square distribution.

    'n' -- The chi-squared parameter; must be one or larger.

    returns -- The probability that a random variable chosen
    according to a chi-square distribution with 'n' degrees of
    freedom will have a value of 'x' or lower."""

    return 1 - ext.prob(x, n)


