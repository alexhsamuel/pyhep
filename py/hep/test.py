#-----------------------------------------------------------------------
#
# module hep.test
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Regression testing support."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.bool import *
import sys

#-----------------------------------------------------------------------
# exceptions
#-----------------------------------------------------------------------

class TestFailure(Exception):
    pass



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def assert_(expression):
    if not expression:
        raise TestFailure, "assertion failed"


def compare(value, expected, precision=None):
    if type(value) in (list, tuple) and type(expected) in (list, tuple):
        if len(value) != len(expected):
            result = False
        else:
            try:
                result = reduce(
                    lambda r1, r2: r1 and r2,
                    map(lambda (v, e): compare(v, e, precision),
                        zip(value, expected)),
                    True)
            except TestFailure:
                result = False

    elif precision is not None:
        result = (abs(value - expected) <= precision)

    else:
        result = (value == expected)

    if result:
        return True
    else:
        raise TestFailure, \
              "expected %s; got %s" % (repr(expected), repr(value))


def compareSequence(value, expected, ordered=True):
    value = list(value)
    expected = list(expected)
    if not ordered:
        value.sort()
        expected.sort()
    return compare(value, expected)


