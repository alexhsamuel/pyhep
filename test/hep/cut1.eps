%!PS-Adobe-1.0 EPSF-1.0
%%BoundingBox: 0 0 400 283
true setstrokeadjust
%% Coordinates are specified in meters.
2834.645669 2834.645669 scale
/ISOEncode { dup length dict begin {1 index /FID ne {def}
{pop pop} ifelse} forall /Encoding ISOLatin1Encoding def
currentdict end /Temporary exch definefont } bind def
0.800000 0.800000 0.800000 setrgbcolor
0.000088 setlinewidth
newpath
0.017000 0.017000 moveto
0.134400 0.017000 lineto
stroke
0.160000 0.220000 0.400000 setrgbcolor
0.000265 setlinewidth
newpath
0.017000 0.017000 moveto
0.017000 0.017000 lineto
stroke
newpath
0.017000 0.017000 moveto
0.019609 0.017000 lineto
stroke
newpath
0.019609 0.017000 moveto
0.019609 0.017084 lineto
stroke
newpath
0.019609 0.017084 moveto
0.022218 0.017084 lineto
stroke
newpath
0.022218 0.017084 moveto
0.022218 0.017084 lineto
stroke
newpath
0.022218 0.017084 moveto
0.024827 0.017084 lineto
stroke
newpath
0.024827 0.017084 moveto
0.024827 0.017252 lineto
stroke
newpath
0.024827 0.017252 moveto
0.027436 0.017252 lineto
stroke
newpath
0.027436 0.017252 moveto
0.027436 0.017503 lineto
stroke
newpath
0.027436 0.017503 moveto
0.030044 0.017503 lineto
stroke
newpath
0.030044 0.017503 moveto
0.030044 0.018510 lineto
stroke
newpath
0.030044 0.018510 moveto
0.032653 0.018510 lineto
stroke
newpath
0.032653 0.018510 moveto
0.032653 0.018846 lineto
stroke
newpath
0.032653 0.018846 moveto
0.035262 0.018846 lineto
stroke
newpath
0.035262 0.018846 moveto
0.035262 0.019601 lineto
stroke
newpath
0.035262 0.019601 moveto
0.037871 0.019601 lineto
stroke
newpath
0.037871 0.019601 moveto
0.037871 0.022034 lineto
stroke
newpath
0.037871 0.022034 moveto
0.040480 0.022034 lineto
stroke
newpath
0.040480 0.022034 moveto
0.040480 0.023544 lineto
stroke
newpath
0.040480 0.023544 moveto
0.043089 0.023544 lineto
stroke
newpath
0.043089 0.023544 moveto
0.043089 0.027738 lineto
stroke
newpath
0.043089 0.027738 moveto
0.045698 0.027738 lineto
stroke
newpath
0.045698 0.027738 moveto
0.045698 0.032101 lineto
stroke
newpath
0.045698 0.032101 moveto
0.048307 0.032101 lineto
stroke
newpath
0.048307 0.032101 moveto
0.048307 0.037974 lineto
stroke
newpath
0.048307 0.037974 moveto
0.050916 0.037974 lineto
stroke
newpath
0.050916 0.037974 moveto
0.050916 0.045105 lineto
stroke
newpath
0.050916 0.045105 moveto
0.053524 0.045105 lineto
stroke
newpath
0.053524 0.045105 moveto
0.053524 0.054249 lineto
stroke
newpath
0.053524 0.054249 moveto
0.056133 0.054249 lineto
stroke
newpath
0.056133 0.054249 moveto
0.056133 0.062219 lineto
stroke
newpath
0.056133 0.062219 moveto
0.058742 0.062219 lineto
stroke
newpath
0.058742 0.062219 moveto
0.058742 0.069937 lineto
stroke
newpath
0.058742 0.069937 moveto
0.061351 0.069937 lineto
stroke
newpath
0.061351 0.069937 moveto
0.061351 0.079921 lineto
stroke
newpath
0.061351 0.079921 moveto
0.063960 0.079921 lineto
stroke
newpath
0.063960 0.079921 moveto
0.063960 0.078746 lineto
stroke
newpath
0.063960 0.078746 moveto
0.066569 0.078746 lineto
stroke
newpath
0.066569 0.078746 moveto
0.066569 0.082521 lineto
stroke
newpath
0.066569 0.082521 moveto
0.069178 0.082521 lineto
stroke
newpath
0.069178 0.082521 moveto
0.069178 0.084199 lineto
stroke
newpath
0.069178 0.084199 moveto
0.071787 0.084199 lineto
stroke
newpath
0.071787 0.084199 moveto
0.071787 0.080592 lineto
stroke
newpath
0.071787 0.080592 moveto
0.074396 0.080592 lineto
stroke
newpath
0.074396 0.080592 moveto
0.074396 0.077404 lineto
stroke
newpath
0.074396 0.077404 moveto
0.077004 0.077404 lineto
stroke
newpath
0.077004 0.077404 moveto
0.077004 0.068427 lineto
stroke
newpath
0.077004 0.068427 moveto
0.079613 0.068427 lineto
stroke
newpath
0.079613 0.068427 moveto
0.079613 0.060541 lineto
stroke
newpath
0.079613 0.060541 moveto
0.082222 0.060541 lineto
stroke
newpath
0.082222 0.060541 moveto
0.082222 0.052739 lineto
stroke
newpath
0.082222 0.052739 moveto
0.084831 0.052739 lineto
stroke
newpath
0.084831 0.052739 moveto
0.084831 0.044685 lineto
stroke
newpath
0.084831 0.044685 moveto
0.087440 0.044685 lineto
stroke
newpath
0.087440 0.044685 moveto
0.087440 0.038980 lineto
stroke
newpath
0.087440 0.038980 moveto
0.090049 0.038980 lineto
stroke
newpath
0.090049 0.038980 moveto
0.090049 0.033695 lineto
stroke
newpath
0.090049 0.033695 moveto
0.092658 0.033695 lineto
stroke
newpath
0.092658 0.033695 moveto
0.092658 0.029332 lineto
stroke
newpath
0.092658 0.029332 moveto
0.095267 0.029332 lineto
stroke
newpath
0.095267 0.029332 moveto
0.095267 0.024131 lineto
stroke
newpath
0.095267 0.024131 moveto
0.097876 0.024131 lineto
stroke
newpath
0.097876 0.024131 moveto
0.097876 0.021698 lineto
stroke
newpath
0.097876 0.021698 moveto
0.100484 0.021698 lineto
stroke
newpath
0.100484 0.021698 moveto
0.100484 0.021027 lineto
stroke
newpath
0.100484 0.021027 moveto
0.103093 0.021027 lineto
stroke
newpath
0.103093 0.021027 moveto
0.103093 0.018678 lineto
stroke
newpath
0.103093 0.018678 moveto
0.105702 0.018678 lineto
stroke
newpath
0.105702 0.018678 moveto
0.105702 0.018007 lineto
stroke
newpath
0.105702 0.018007 moveto
0.108311 0.018007 lineto
stroke
newpath
0.108311 0.018007 moveto
0.108311 0.017587 lineto
stroke
newpath
0.108311 0.017587 moveto
0.110920 0.017587 lineto
stroke
newpath
0.110920 0.017587 moveto
0.110920 0.017252 lineto
stroke
newpath
0.110920 0.017252 moveto
0.113529 0.017252 lineto
stroke
newpath
0.113529 0.017252 moveto
0.113529 0.017000 lineto
stroke
newpath
0.113529 0.017000 moveto
0.116138 0.017000 lineto
stroke
newpath
0.116138 0.017000 moveto
0.116138 0.017000 lineto
stroke
newpath
0.116138 0.017000 moveto
0.118747 0.017000 lineto
stroke
newpath
0.118747 0.017000 moveto
0.118747 0.017000 lineto
stroke
newpath
0.118747 0.017000 moveto
0.121356 0.017000 lineto
stroke
newpath
0.121356 0.017000 moveto
0.121356 0.017000 lineto
stroke
newpath
0.121356 0.017000 moveto
0.123964 0.017000 lineto
stroke
newpath
0.123964 0.017000 moveto
0.123964 0.017000 lineto
stroke
newpath
0.123964 0.017000 moveto
0.126573 0.017000 lineto
stroke
newpath
0.126573 0.017000 moveto
0.126573 0.017000 lineto
stroke
newpath
0.126573 0.017000 moveto
0.129182 0.017000 lineto
stroke
newpath
0.129182 0.017000 moveto
0.129182 0.017000 lineto
stroke
newpath
0.129182 0.017000 moveto
0.131791 0.017000 lineto
stroke
newpath
0.131791 0.017000 moveto
0.131791 0.017000 lineto
stroke
newpath
0.131791 0.017000 moveto
0.134400 0.017000 lineto
stroke
newpath
0.134400 0.017000 moveto
0.134400 0.017000 lineto
stroke
0.400000 0.160000 0.340000 setrgbcolor
newpath
0.017000 0.017000 moveto
0.017000 0.017839 lineto
stroke
newpath
0.017000 0.017839 moveto
0.019609 0.017839 lineto
stroke
newpath
0.019609 0.017839 moveto
0.019609 0.018426 lineto
stroke
newpath
0.019609 0.018426 moveto
0.022218 0.018426 lineto
stroke
newpath
0.022218 0.018426 moveto
0.022218 0.021111 lineto
stroke
newpath
0.022218 0.021111 moveto
0.024827 0.021111 lineto
stroke
newpath
0.024827 0.021111 moveto
0.024827 0.021111 lineto
stroke
newpath
0.024827 0.021111 moveto
0.027436 0.021111 lineto
stroke
newpath
0.027436 0.021111 moveto
0.027436 0.024215 lineto
stroke
newpath
0.027436 0.024215 moveto
0.030044 0.024215 lineto
stroke
newpath
0.030044 0.024215 moveto
0.030044 0.028074 lineto
stroke
newpath
0.030044 0.028074 moveto
0.032653 0.028074 lineto
stroke
newpath
0.032653 0.028074 moveto
0.032653 0.034618 lineto
stroke
newpath
0.032653 0.034618 moveto
0.035262 0.034618 lineto
stroke
newpath
0.035262 0.034618 moveto
0.035262 0.038896 lineto
stroke
newpath
0.035262 0.038896 moveto
0.037871 0.038896 lineto
stroke
newpath
0.037871 0.038896 moveto
0.037871 0.044433 lineto
stroke
newpath
0.037871 0.044433 moveto
0.040480 0.044433 lineto
stroke
newpath
0.040480 0.044433 moveto
0.040480 0.054165 lineto
stroke
newpath
0.040480 0.054165 moveto
0.043089 0.054165 lineto
stroke
newpath
0.043089 0.054165 moveto
0.043089 0.061548 lineto
stroke
newpath
0.043089 0.061548 moveto
0.045698 0.061548 lineto
stroke
newpath
0.045698 0.061548 moveto
0.045698 0.071279 lineto
stroke
newpath
0.045698 0.071279 moveto
0.048307 0.071279 lineto
stroke
newpath
0.048307 0.071279 moveto
0.048307 0.077739 lineto
stroke
newpath
0.048307 0.077739 moveto
0.050916 0.077739 lineto
stroke
newpath
0.050916 0.077739 moveto
0.050916 0.081431 lineto
stroke
newpath
0.050916 0.081431 moveto
0.053524 0.081431 lineto
stroke
newpath
0.053524 0.081431 moveto
0.053524 0.086297 lineto
stroke
newpath
0.053524 0.086297 moveto
0.056133 0.086297 lineto
stroke
newpath
0.056133 0.086297 moveto
0.056133 0.082186 lineto
stroke
newpath
0.056133 0.082186 moveto
0.058742 0.082186 lineto
stroke
newpath
0.058742 0.082186 moveto
0.058742 0.078075 lineto
stroke
newpath
0.058742 0.078075 moveto
0.061351 0.078075 lineto
stroke
newpath
0.061351 0.078075 moveto
0.061351 0.075558 lineto
stroke
newpath
0.061351 0.075558 moveto
0.063960 0.075558 lineto
stroke
newpath
0.063960 0.075558 moveto
0.063960 0.067085 lineto
stroke
newpath
0.063960 0.067085 moveto
0.066569 0.067085 lineto
stroke
newpath
0.066569 0.067085 moveto
0.066569 0.062806 lineto
stroke
newpath
0.066569 0.062806 moveto
0.069178 0.062806 lineto
stroke
newpath
0.069178 0.062806 moveto
0.069178 0.054333 lineto
stroke
newpath
0.069178 0.054333 moveto
0.071787 0.054333 lineto
stroke
newpath
0.071787 0.054333 moveto
0.071787 0.043427 lineto
stroke
newpath
0.071787 0.043427 moveto
0.074396 0.043427 lineto
stroke
newpath
0.074396 0.043427 moveto
0.074396 0.038645 lineto
stroke
newpath
0.074396 0.038645 moveto
0.077004 0.038645 lineto
stroke
newpath
0.077004 0.038645 moveto
0.077004 0.031849 lineto
stroke
newpath
0.077004 0.031849 moveto
0.079613 0.031849 lineto
stroke
newpath
0.079613 0.031849 moveto
0.079613 0.028661 lineto
stroke
newpath
0.079613 0.028661 moveto
0.082222 0.028661 lineto
stroke
newpath
0.082222 0.028661 moveto
0.082222 0.024550 lineto
stroke
newpath
0.082222 0.024550 moveto
0.084831 0.024550 lineto
stroke
newpath
0.084831 0.024550 moveto
0.084831 0.021698 lineto
stroke
newpath
0.084831 0.021698 moveto
0.087440 0.021698 lineto
stroke
newpath
0.087440 0.021698 moveto
0.087440 0.019181 lineto
stroke
newpath
0.087440 0.019181 moveto
0.090049 0.019181 lineto
stroke
newpath
0.090049 0.019181 moveto
0.090049 0.018762 lineto
stroke
newpath
0.090049 0.018762 moveto
0.092658 0.018762 lineto
stroke
newpath
0.092658 0.018762 moveto
0.092658 0.018175 lineto
stroke
newpath
0.092658 0.018175 moveto
0.095267 0.018175 lineto
stroke
newpath
0.095267 0.018175 moveto
0.095267 0.017419 lineto
stroke
newpath
0.095267 0.017419 moveto
0.097876 0.017419 lineto
stroke
newpath
0.097876 0.017419 moveto
0.097876 0.017587 lineto
stroke
newpath
0.097876 0.017587 moveto
0.100484 0.017587 lineto
stroke
newpath
0.100484 0.017587 moveto
0.100484 0.017252 lineto
stroke
newpath
0.100484 0.017252 moveto
0.103093 0.017252 lineto
stroke
newpath
0.103093 0.017252 moveto
0.103093 0.017084 lineto
stroke
newpath
0.103093 0.017084 moveto
0.105702 0.017084 lineto
stroke
newpath
0.105702 0.017084 moveto
0.105702 0.017000 lineto
stroke
newpath
0.105702 0.017000 moveto
0.108311 0.017000 lineto
stroke
newpath
0.108311 0.017000 moveto
0.108311 0.017000 lineto
stroke
newpath
0.108311 0.017000 moveto
0.110920 0.017000 lineto
stroke
newpath
0.110920 0.017000 moveto
0.110920 0.017000 lineto
stroke
newpath
0.110920 0.017000 moveto
0.113529 0.017000 lineto
stroke
newpath
0.113529 0.017000 moveto
0.113529 0.017000 lineto
stroke
newpath
0.113529 0.017000 moveto
0.116138 0.017000 lineto
stroke
newpath
0.116138 0.017000 moveto
0.116138 0.017000 lineto
stroke
newpath
0.116138 0.017000 moveto
0.118747 0.017000 lineto
stroke
newpath
0.118747 0.017000 moveto
0.118747 0.017000 lineto
stroke
newpath
0.118747 0.017000 moveto
0.121356 0.017000 lineto
stroke
newpath
0.121356 0.017000 moveto
0.121356 0.017000 lineto
stroke
newpath
0.121356 0.017000 moveto
0.123964 0.017000 lineto
stroke
newpath
0.123964 0.017000 moveto
0.123964 0.017000 lineto
stroke
newpath
0.123964 0.017000 moveto
0.126573 0.017000 lineto
stroke
newpath
0.126573 0.017000 moveto
0.126573 0.017000 lineto
stroke
newpath
0.126573 0.017000 moveto
0.129182 0.017000 lineto
stroke
newpath
0.129182 0.017000 moveto
0.129182 0.017000 lineto
stroke
newpath
0.129182 0.017000 moveto
0.131791 0.017000 lineto
stroke
newpath
0.131791 0.017000 moveto
0.131791 0.017000 lineto
stroke
newpath
0.131791 0.017000 moveto
0.134400 0.017000 lineto
stroke
newpath
0.134400 0.017000 moveto
0.134400 0.017000 lineto
stroke
/Times-Roman findfont 0.003352 scalefont ISOEncode setfont
0.000000 0.000000 0.000000 setrgbcolor
0.017000 0.013162 moveto
(-4)
dup stringwidth pop -0.500000 mul -0.002514 rmoveto
show
0.000035 setlinewidth
newpath
0.017000 0.015000 moveto
0.017000 0.014000 lineto
stroke
0.043089 0.013162 moveto
(-2)
dup stringwidth pop -0.500000 mul -0.002514 rmoveto
show
newpath
0.043089 0.015000 moveto
0.043089 0.014000 lineto
stroke
0.069178 0.013162 moveto
(0)
dup stringwidth pop -0.500000 mul -0.002514 rmoveto
show
newpath
0.069178 0.015000 moveto
0.069178 0.014000 lineto
stroke
0.095267 0.013162 moveto
(2)
dup stringwidth pop -0.500000 mul -0.002514 rmoveto
show
newpath
0.095267 0.015000 moveto
0.095267 0.014000 lineto
stroke
0.121356 0.013162 moveto
(4)
dup stringwidth pop -0.500000 mul -0.002514 rmoveto
show
newpath
0.121356 0.015000 moveto
0.121356 0.014000 lineto
stroke
0.013162 0.017000 moveto
(0)
dup stringwidth pop -1.000000 mul -0.001257 rmoveto
show
newpath
0.014000 0.017000 moveto
0.015000 0.017000 lineto
stroke
0.013162 0.033779 moveto
(200)
dup stringwidth pop -1.000000 mul -0.001257 rmoveto
show
newpath
0.014000 0.033779 moveto
0.015000 0.033779 lineto
stroke
0.013162 0.050558 moveto
(400)
dup stringwidth pop -1.000000 mul -0.001257 rmoveto
show
newpath
0.014000 0.050558 moveto
0.015000 0.050558 lineto
stroke
0.013162 0.067336 moveto
(600)
dup stringwidth pop -1.000000 mul -0.001257 rmoveto
show
newpath
0.014000 0.067336 moveto
0.015000 0.067336 lineto
stroke
0.013162 0.084115 moveto
(800)
dup stringwidth pop -1.000000 mul -0.001257 rmoveto
show
newpath
0.014000 0.084115 moveto
0.015000 0.084115 lineto
stroke
0.015838 0.086297 moveto
/Times findfont 0.003352 scalefont ISOEncode setfont
(entries / 0.2 ) show
0.070700 0.093000 moveto
/Times findfont 0.003352 scalefont ISOEncode setfont
(gauss) show
currentfont
/Times findfont 0.003352 scalefont ISOEncode setfont
(\() show
setfont
currentfont
/Symbol findfont 0.003352 scalefont ISOEncode
[ 1 0 0.15 1 0 0 ] makefont setfont
/mu glyphshow
setfont
currentfont
/Times findfont 0.003352 scalefont ISOEncode setfont
(,) show
setfont
currentfont
/Symbol findfont 0.003352 scalefont ISOEncode
[ 1 0 0.15 1 0 0 ] makefont setfont
/sigma glyphshow
setfont
currentfont
/Times findfont 0.003352 scalefont ISOEncode setfont
(\)) show
setfont
0.160000 0.220000 0.400000 setrgbcolor
0.000265 setlinewidth
newpath
0.104400 0.083783 moveto
0.109400 0.083783 lineto
stroke
0.000000 0.000000 0.000000 setrgbcolor
0.111400 0.082945 moveto
/Times findfont 0.003352 scalefont ISOEncode setfont
currentfont
/Symbol findfont 0.003352 scalefont ISOEncode
[ 1 0 0.15 1 0 0 ] makefont setfont
/mu glyphshow
setfont
currentfont
/Times findfont 0.003352 scalefont ISOEncode setfont
(=) show
setfont
currentfont
/Times findfont 0.003352 scalefont ISOEncode setfont
(0) show
setfont
(, ) show
currentfont
/Symbol findfont 0.003352 scalefont ISOEncode
[ 1 0 0.15 1 0 0 ] makefont setfont
/sigma glyphshow
setfont
currentfont
/Times findfont 0.003352 scalefont ISOEncode setfont
(=) show
setfont
currentfont
/Times findfont 0.003352 scalefont ISOEncode setfont
(1) show
setfont
0.400000 0.160000 0.340000 setrgbcolor
newpath
0.104400 0.078755 moveto
0.109400 0.078755 lineto
stroke
0.000000 0.000000 0.000000 setrgbcolor
0.111400 0.077917 moveto
/Times findfont 0.003352 scalefont ISOEncode setfont
currentfont
/Symbol findfont 0.003352 scalefont ISOEncode
[ 1 0 0.15 1 0 0 ] makefont setfont
/mu glyphshow
setfont
currentfont
/Times findfont 0.003352 scalefont ISOEncode setfont
(=) show
setfont
/minus glyphshow
currentfont
/Times findfont 0.003352 scalefont ISOEncode setfont
(1) show
setfont
(, ) show
currentfont
/Symbol findfont 0.003352 scalefont ISOEncode
[ 1 0 0.15 1 0 0 ] makefont setfont
/sigma glyphshow
setfont
currentfont
/Times findfont 0.003352 scalefont ISOEncode setfont
(=) show
setfont
currentfont
/Times findfont 0.003352 scalefont ISOEncode setfont
(1) show
setfont
0.700000 0.700000 0.700000 setrgbcolor
newpath
0.052860 0.017000 moveto
0.052860 0.086297 lineto
stroke
