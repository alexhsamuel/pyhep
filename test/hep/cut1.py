#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

import hep.cuts
from   hep.num import in_range
from   hep.test import assert_
from   numarray import array, Float32
from   random import normalvariate, random
import os

#-----------------------------------------------------------------------
# test
#-----------------------------------------------------------------------

sig_values = array(shape=(1000, 4), type=Float32)
for i in range(sig_values.shape[0]):
    sig_values[i, 0] = normalvariate(0, 1)
    sig_values[i, 1] = normalvariate(0, 1)
    sig_values[i, 2] = normalvariate(0, 1)
    sig_values[i, 3] = normalvariate(0, 1)

bkg_values = array(shape=(5000, 4), type=Float32)
for i in range(bkg_values.shape[0]):
    bkg_values[i, 0] = normalvariate( 0, 2)
    bkg_values[i, 1] = normalvariate( 1, 1)
    bkg_values[i, 2] = normalvariate(-1, 1)
    bkg_values[i, 3] = normalvariate(-1, 1)

cuts = [
    (0, "<", random()),
    (0, ">", random()),
    (1, "<", random()),
    (2, ">", random()),
    (3, ">", random()),
    ]    

cuts, fom = hep.cuts.iterativeOptimize(sig_values, bkg_values, cuts)

assert_(in_range(0, cuts[0][2], 3))
assert_(in_range(-3, cuts[1][2], 0))
assert_(cuts[2][2] > 0)
assert_(cuts[3][2] < 0)
assert_(cuts[4][2] < 0)

assert_(fom > 300)
