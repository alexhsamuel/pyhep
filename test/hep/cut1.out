Traceback (most recent call last):
  File "cut1.py", line 5, in ?
    import hep.cuts
  File "/home/samuel/dev/pyhep/py/hep/cuts.py", line 20, in ?
    from   numarray import *
ImportError: No module named numarray
